staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/implementation.go

```diff
@@ -0,0 +1,314 @@
+	return c.Matcher.Matches(&criteria, a, o)
+}
+
+func (c *CValidatorCompiler) ValidateInitialization() error {
+	if err := c.Matcher.ValidateInitialization(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+type validationActivation struct {
+	object    *unstructured.Unstructured
+	oldObject *unstructured.Unstructured
+	params    *unstructured.Unstructured
+	request   *unstructured.Unstructured
+	hasOld    bool
+}
+
+func (a *validationActivation) ResolveName(name string) (interface{}, bool) {
+	switch name {
+	case ObjectVarName:
+		return a.object.Object, true
+	case OldObjectVarName:
+		return a.oldObject.Object, a.hasOld
+	case ParamsVarName:
+		return a.params.Object, true
+	case RequestVarName:
+		return a.request.Object, true
+	default:
+		return nil, false
+	}
+}
+
+func (a *validationActivation) Parent() interpreter.Activation {
+	return nil
+}
+
+func (c *CValidatorCompiler) Compile(p *v1alpha1.ValidatingAdmissionPolicy) (Validator, error) {
+	if len(p.Spec.Validations) == 0 {
+		return nil, nil
+	}
+	hasParam := false
+	if p.Spec.ParamKind != nil {
+		hasParam = true
+	}
+	compilationResults := make([]CompilationResult, len(p.Spec.Validations))
+	for i, validation := range p.Spec.Validations {
+		compilationResults[i] = CompileValidatingPolicyExpression(validation.Expression, hasParam)
+	}
+	return &CValidator{policy: p, compilationResults: compilationResults}, nil
+}
+
+type CValidator struct {
+	policy                *v1alpha1.ValidatingAdmissionPolicy
+	compilationResults    []CompilationResult
+	getOpenAPIDefinitions common.GetOpenAPIDefinitions
+}
+
+func convertObjectToUnstructured(obj interface{}) *unstructured.Unstructured {
+	var unstructuredObj unstructured.Unstructured
+	jsonObj, _ := json.Marshal(obj)
+	_ = json.Unmarshal(jsonObj, &unstructuredObj.Object)
+
+	return &unstructuredObj
+}
+
+func (v *CValidator) Validate(a admission.Attributes, o admission.ObjectInterfaces, params *unstructured.Unstructured) ([]PolicyDecision, error) {
+	decisions := make([]PolicyDecision, len(v.policy.Spec.Validations))
+
+	/* TODO: implement schema-based validation.
+	refFunc := func(name string) spec.Ref {
+		return spec.MustCreateRef(fmt.Sprintf("#/definitions/%s", friendlyName(name)))
+	}
+	defs := v.getOpenAPIDefinitions(refFunc)
+
+	objectVal, err := convertObjectToVal(a.GetObject(), defs, "")
+	if err != nil {
+		return nil, err
+	}
+
+	hasOld := false
+	var oldObjectVal ref.Val = types.NullValue
+	if a.GetOldObject() != nil {
+		oldObjectVal, err = convertObjectToVal(a.GetOldObject(), defs, "")
+		if err != nil {
+			return nil, err
+		}
+		hasOld = true
+	}
+
+	requestVal, err := convertObjectToVal(createAdmissionRequest(a), defs, admissionRequesGVK)
+	*/
+
+	hasOld := a.GetOldObject() != nil
+	var oldObjectVal *unstructured.Unstructured = nil
+	if hasOld {
+		oldObjectVal = convertObjectToUnstructured(a.GetOldObject())
+	}
+
+	va := &validationActivation{
+		object:    convertObjectToUnstructured(a.GetObject()),
```

Comment:
note that object might be nil (`ToUnstructured` will return an error if a nil passed in). In CEL is should be possible to check `object == null" (which will always be true on DELETE)